File: go/cote/GJM/최대증가부분수열.go
package main

import (
	"fmt"
	"math/rand"
)

const (
	TEST_CASES = 10
	N = 20
	MAX_VALUE = 100
)


var cache = make([]int, N+1)

func max(x,y int) int {
	if x < y {
		return y
	}
	return x 
} 

func search(cur int, l []int) int {

	var ret *int

	ret = &cache[cur+1]

	if *ret != -1 {
		return *ret
	}

	*ret = 0

	for i := cur+1; i < len(l); i++ {
		if cur == -1 || l[cur] < l[i] {
			*ret = max(*ret, 1 + search(i, l))
		}
	}

	return *ret
}

func solve(l []int) int {

	for i, _ := range cache {
		cache[i] = -1
	}

	return search(-1, l)
}

func main() {
	rand.Seed(43)
	for it := 0; it < TEST_CASES; it++ {
		n := rand.Intn(N-10)+10
		l := make([]int, n)

		for i := 0; i < n; i++ {
			l[i] = rand.Intn(MAX_VALUE)
		}
		fmt.Println(solve(l))
	}
}

File: python/cote/GJM/최대증가부분수열.py
import random

N = 20
MAX_VALUE = 100
TEST_CASES = 10

cache = [-1] * (N + 1)
answer = [-1] * (N + 1)

random.seed(43)


def solve(cur, l):

    ret = cache[cur + 1]

    if ret != -1:
        return ret

    ret = 1
    j = -1

    for i in range(0, N):
        if cur == -1 or l[cur] < l[i]:
            tmp = 1 + solve(i, l)
            if ret < tmp:
                ret = tmp
                j = i

    cache[cur + 1] = ret
    answer[cur + 1] = j

    return ret


def print_answer(cur, l, v):

    if cur != -1:
        v.append(l[cur])

    next_idx = answer[cur + 1]

    if next_idx != -1:
        print_answer(next_idx, l, v)


if __name__ == "__main__":

    for _ in range(0, TEST_CASES):

        cache = [-1] * (N + 1)
        answer = [-1] * (N + 1)
        l = [random.randint(0, MAX_VALUE) for _ in range(0, N)]
        v = []
        print(solve(-1, l) - 1)
        print_answer(-1, l, v)
        print(v)
